File: dsa/array/sorting/counting_sort.js
/* 
Stable : No

Time Complexity: O(n+k) 
    - n = number of elements in the input array
    - k = maximum element value
    - Counting sort is linear if k is not much bigger than n, otherwise
    time depends on the range of values.
    
Space Complexity: O(k)
    - Array is modified in place.
*/
function count_sort(arr) {
    // Use max function to find maximum value in given array.
    let max_val = Math.max(...arr);

    // Initialize count array for storing the values  of given array.
    let count = new Array(max_val + 1).fill(0);

    /* Destroys the original array by popping elements one by one
    and storing them in earlier initialized count array. */
    while (arr.length > 0) {
        // Removes element one by one.
        let num = arr.pop();
        // Assigns the popped value to count array index.
        count[num] += 1;
    }

    /* Rebuilds the original array by taking elements from the 
    count array, stored as its index.*/
    for (let i = 0; i < count.length; i++) {
        while (count[i] > 0) {
            // Adds the index of count array, i.e., original element to array.
            // --- Indexes are already sorted for each iteration ---
            arr.push(i);
            // removes the value at index from count array.
            count[i] -= 1;
        }
    }

    // Returns the sorted array.
    return arr;
}


// ---- Examples ---- 

const example1 = [38, 70, 39, 1, 5, 87, 13];
const example2 = [5, 2, 3, 0, 1];
const example3 = [99, 110, 18, 4, 6, 63];
const example4 = [];

console.log("Example 1:");
console.log("Input:", example1);
console.log("Sorted:", count_sort(example1));

console.log("\nExample 2:");
console.log("Input:", example2);
console.log("Sorted:", count_sort(example2));

console.log("\nExample 3:");
console.log("Input:", example3);
console.log("Sorted:", count_sort(example3));

console.log("Example 4:");
console.log("Input:", example4);
console.log("Sorted:", count_sort(example4));

File: dsa_in_java/sorting/CountingSort.java
/*
 * Topic: Counting Sort
 * Language: Java
 *
 * Description:
 * Counting Sort is a non-comparison-based algorithm suitable for integers within a known range.
 * It counts occurrences of each element and uses that information to build the sorted array.
 *
 * Time Complexity:
 * O(n + k), where k = range of input.
 *
 * Space Complexity:
 * O(n + k)
 */

public class CountingSort {

    public static void countingSort(int[] arr) {
        int max = arr[0];
        for (int num : arr)
            if (num > max) max = num;

        int[] count = new int[max + 1];

        // Count occurrences
        for (int num : arr)
            count[num]++;

        // Reconstruct sorted array
        int index = 0;
        for (int i = 0; i <= max; i++) {
            while (count[i] > 0) {
                arr[index++] = i;
                count[i]--;
            }
        }
    }

    public static void printArray(int[] arr) {
        for (int num : arr)
            System.out.print(num + " ");
        System.out.println();
    }

    public static void main(String[] args) {
        int[] arr = {4, 2, 2, 8, 3, 3, 1};
        System.out.println("Original array:");
        printArray(arr);

        countingSort(arr);

        System.out.println("Sorted array:");
        printArray(arr);
    }
}


/*
 * Example Output:
 * ------------------------------
 * Original array:
 * 4 2 2 8 3 3 1
 * Sorted array:
 * 1 2 2 3 3 4 8
 */